mod/primitives/pkg/encoding/ssz/schema/schema.go

```diff
@@ -0,0 +1,205 @@
+}
+
+// Basic Type
+
+type basic struct {
+	size uint64
+}
+
+func UInt8() SSZType   { return basic{size: 1} }
+func UInt16() SSZType  { return basic{size: 2} }
+func UInt32() SSZType  { return basic{size: 4} }
+func UInt64() SSZType  { return basic{size: 8} }
+func UInt128() SSZType { return basic{size: 16} }
+func UInt256() SSZType { return basic{size: 32} }
+
+func (b basic) Size() uint64 { return b.size }
+
+func (b basic) Chunks() uint64 { return 1 }
+
+func (b basic) child(_ string) SSZType { return b }
+
+func (b basic) position(_ string) (uint64, uint8, error) {
+	return 0, 0, errors.New("basic type has no children")
+}
+
+// Container Type
+
+type container struct {
+	Fields     []SSZType
+	FieldIndex map[string]uint64
+}
+
+type field struct {
+	name string
+	typ  SSZType
+}
+
+func Field(name string, typ SSZType) field {
+	return field{name: name, typ: typ}
+}
+
+func Container(fields ...field) SSZType {
+	fieldIndex := make(map[string]uint64)
+	types := make([]SSZType, len(fields))
+	for i, f := range fields {
+		fieldIndex[f.name] = uint64(i)
+		types[i] = f.typ
+	}
+	return container{Fields: types, FieldIndex: fieldIndex}
+}
+
+func (c container) Size() uint64 { return chunkSize }
+
+func (c container) Length() uint64 { return uint64(len(c.Fields)) }
+
+func (c container) Chunks() uint64 { return uint64(len(c.Fields)) }
+
+func (c container) child(p string) SSZType {
+	return c.Fields[c.FieldIndex[p]]
+}
+
+func (c container) position(p string) (uint64, uint8, error) {
+	pos, ok := c.FieldIndex[p]
+	if !ok {
+		return 0, 0, fmt.Errorf("field %s not found", p)
+	}
+	return pos, 0, nil
+}
+
+// Enumerable Type (vectors and lists)
+
+func List(element SSZType, length uint64) SSZType {
+	return enumerable{Element: element, maxLength: length}
+}
+
+func Vector(element SSZType, length uint64) SSZType {
+	return enumerable{Element: element, length: length}
+}
+
+func Bytes(length uint64) SSZType {
+	return Vector(UInt8(), length)
+}
+
+type enumerable struct {
+	Element   SSZType
+	length    uint64
+	maxLength uint64
+}
+
+func (e enumerable) Size() uint64 { return chunkSize }
+
+func (e enumerable) Chunks() uint64 {
+	x := float64(e.Length()*e.Element.Size()) / chunkSize
+	return uint64(math.Ceil(x))
+}
+
+func (e enumerable) child(_ string) SSZType {
+	return e.Element
+}
+
+func (e enumerable) Length() uint64 {
+	if e.length == 0 {
+		return e.maxLength
+	}
+	return e.length
+}
+
+func (e enumerable) position(p string) (uint64, uint8, error) {
+	i, err := strconv.ParseUint(p, 10, 64)
+	if err != nil {
+		return 0, 0, fmt.Errorf("expected index, got name %s", p)
+	}
+	start := i * e.Element.Size()
+	return uint64(math.Floor(float64(start) / chunkSize)),
+		uint8(start % chunkSize),
+		nil
+}
+
+func (e enumerable) IsByteVector() bool {
+	return e.Element.Size() == 1 && e.length > 0
+}
+
+func (e enumerable) IsList() bool {
+	return e.maxLength > 0
+}
+
+func (e enumerable) IsFixed() bool {
+	_, ok := e.Element.(basic)
```

Comment:
this is wrong, what is the element is a Vector of ints

Reply:
what are you seeing? I added a test for a vector of 128 bit ints, I think it's correct.

Reply:
oh, well from the spec:
>We recursively define "variable-size" types to be lists, unions, Bitlist and all types that contain a variable-size type.
so to determine if a vector is fixed length the entire subtree should be recursed for any var sized type? I've removed it for now since I don't think there's a use case from the schema side.